packages/aws-cdk/lib/docker.ts

```diff
@@ -0,0 +1,208 @@
+import { debug, print } from './logging';
+import { shell } from './os';
+import { PleaseHold } from './util/please-hold';
+
+/**
+ * Build and upload a Docker image
+ *
+ * Permanently identifying images is a bit of a bust. Newer Docker version use
+ * a digest (sha256:xxxx) as an image identifier, which is pretty good to avoid
+ * spurious rebuilds. However, this digest is calculated over a manifest that
+ * includes metadata that is liable to change. For example, as soon as we
+ * push the Docker image to a repository, the digest changes. This makes the
+ * digest worthless to determe whether we already pushed an image, for example.
+ *
+ * As a workaround, we calculate our own digest over parts of the manifest that
+ * are unlikely to change, and tag based on that.
+ */
+export async function prepareContainerAsset(asset: ContainerImageAssetMetadataEntry, toolkitInfo: ToolkitInfo): Promise<CloudFormation.Parameter[]> {
+  debug(' 👑  Preparing Docker image asset:', asset.path);
+
+  const buildHold = new PleaseHold(` ⌛ Building Docker image for ${asset.path}; this may take a while.`);
+  try {
+    buildHold.start();
+
+    const command = ['docker',
+      'build',
+      '--quiet',
+      asset.path];
+    const imageId = (await shell(command, { quiet: true })).trim();
+    buildHold.stop();
+
+    const tag = await calculateImageDigest(imageId);
+
+    debug(` ⌛  Image has tag ${tag}, preparing ECR repository`);
+    const ecr = await toolkitInfo.prepareEcrRepository(asset.id, tag);
+
+    if (ecr.alreadyExists) {
+      debug(' 👑  Image already uploaded.');
+    } else {
+      // Login and push
+      debug(` ⌛  Image needs to be uploaded first.`);
+
+      await shell(['docker', 'login',
+        '--username', ecr.username,
+        '--password', ecr.password,
+        ecr.endpoint]);
+
+      const qualifiedImageName = `${ecr.repositoryUri}:${tag}`;
+      await shell(['docker', 'tag', imageId, qualifiedImageName]);
+
+      // There's no way to make this quiet, so we can't use a PleaseHold. Print a header message.
+      print(` ⌛ Pusing Docker image for ${asset.path}; this may take a while.`);
+      await shell(['docker', 'push', qualifiedImageName]);
+    }
+
+    return [
+      { ParameterKey: asset.repositoryParameter, ParameterValue: ecr.repositoryArn },
+      { ParameterKey: asset.tagParameter, ParameterValue: tag },
+    ];
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      // tslint:disable-next-line:max-line-length
+      throw new Error('Error building Docker image asset; you need to have Docker installed in order to be able to build image assets. Please install Docker and try again.');
+    }
+    throw e;
+  } finally {
+    buildHold.stop();
+  }
+}
+
+/**
+ * Calculate image digest
+ */
+async function calculateImageDigest(imageId: string) {
+  const manifest = await shell(['docker', 'inspect', imageId], { quiet: true });
+  const parsed = JSON.parse(manifest)[0];
+
+  const importantParts = {
```

Comment:
Isn't this one case where it's safer to build one time too many rather than once too few? In such cases, I would rather apply a black-list of known-irrelevant keys, instead of a white-list of known-stable ones...

Reply:
Tend to agree